core/shared/src/main/scala/zio/ZRef.scala

```diff
@@ -0,0 +1,343 @@
+  //   unified.get.flatMap[Any, E, B](b => f(b)).uninterruptible
+  // }
+
+}
+object ZRef extends Serializable {
+
+  final case class AtomicRef[E, A](value: AtomicReference[A]) extends ZRef[E, E, A, A] { self =>
+    final def get: IO[E, A] = IO.effectTotal(value.get)
+
+    final def set(a: A): IO[E, Unit] = IO.effectTotal(value.set(a))
+
+    final def setAsync(a: A): IO[E, Unit] = IO.effectTotal(value.lazySet(a))
+
+    // override final def maybeUpdate[E2](f: A => Either[E2,A])(implicit ea: E <:< E2, eb: E <:< E2):zio.IO[E2,A] =
+    // UIO.effectTotal(value.updateAndGet(a => f(a).getOrElse(???)))
+
+    override final def update[E1](f: A => A)(implicit ea: E <:< E1, eb: E <:< E1): zio.IO[E1, A] =
+      unsafeUpdate(f.asInstanceOf[Any => Any]).mapError(e => ea.apply(e))
+
+    override final def updateSome[E1](
+      pf: PartialFunction[A, A]
+    )(implicit ea: E <:< E1, eb: E <:< E1): zio.IO[E1, A] = {
+      val f: A => A = a => pf.lift(a).getOrElse(a)
+      self.update(f)(ea, eb)
+    }
+
+    override final def modify[E1, C](f: A => (C, A))(implicit ea: E <:< E1, eb: E <:< E1): zio.IO[E1, C] =
+      unsafeModify(f.asInstanceOf[Any => (Any, Any)])
+
+    override final def modifySome[E1, C](
+      default: C
+    )(pf: PartialFunction[A, (C, A)])(implicit ea: E <:< E1, eb: E <:< E1): zio.IO[E1, C] = {
+      val f: A => (C, A) = a => pf.lift(a).getOrElse((default, a))
+      self.modify(f)(ea, eb)
+    }
+
+    private[this] final def unsafeModify[E, B](f: Any => (Any, Any)): IO[E, B] = IO.effectTotal {
+      var loop = true
+      var b: B = null.asInstanceOf[B]
+      while (loop) {
+        val current = value.get
+
+        val tuple = f(current)
+
+        b = tuple._1.asInstanceOf[B]
+
+        loop = !value.compareAndSet(current, tuple._2.asInstanceOf[A])
+      }
+
+      b
+    }
+
+    private[this] final def unsafeUpdate(f: Any => Any): IO[E, A] = IO.effectTotal {
+      var loop    = true
+      var next: A = null.asInstanceOf[A]
+
+      while (loop) {
+        val current = value.get
+
+        next = f(current).asInstanceOf[A]
+
+        loop = !value.compareAndSet(current, next)
+      }
+      next
+    }
+  }
+
+  final case class Derived[EA0, EB0, A0, B0, EA, EB, A, B](
+    original: ZRef[EA0, EB0, A0, B0],
+    getter: Either[EB0, B0] => Either[EB, B],
+    setter: Either[EA0, A] => Either[EA, A0]
+  )(implicit fea: EA0 => EA)
+      extends ZRef[EA, EB, A, B] { self =>
+    def get: zio.IO[EB, B] =
+      original.get.either.flatMap(e => ZIO.fromEither(getter(e)))
+    def set(a: A): zio.IO[EA, Unit] = ZIO.fromEither(setter(Right(a))).flatMap(a0 => original.set(a0).mapError(fea))
+
+    def setAsync(a: A): zio.IO[EA, Unit] = ZIO.fromEither(setter(Right(a))).map(a0 => original.setAsync(a0)).unit
+
+    def modify[E, C](f: B => (C, A))(implicit ea: EA <:< E, eb: EB <:< E): zio.IO[E, C] = get.mapError(eb.apply).flatMap{ b =>
```

Comment:
That seems too easy, it's probably wrong, I need guidance here @jdegoes 🙏

Reply:
This unfortunately will not be atomic. Any time you separate `get` from `set`, i.e. do state changes outside `update` or `modify`, it will not have atomic guarantees. Gotta stitch together the `Derived` functions all the way back to the root, to create "big getter" and big setter", which can be fused into a function that can be passed to the `AtomicRef`'s update / modify.

Reply:
@dsebban You can apply the same reasoning here, i.e. the only thing you can do is use `original`:
```scala
def modify[E, C](f: B => (C, A))(implicit ea: EA <:< E, eb: EB <:< E): zio.IO[E, C] =
  original.modify(???)
```
We need to provide a function `B0 => (C, A0)` however as you remember failure may occur at some point, so we need:
```scala
private def modifyOrFail[E, C](f: B => Either[E, (C, A)])(implicit ea: EA <:< E, eb: EB <:< E): zio.IO[E, C] =
  original.modifyOrFail(???)

def modify[E, C](f: B => (C, A))(implicit ea: EA <:< E, eb: EB <:< E): zio.IO[E, C] =
  modifyOrFail(b => Right(f(b)))
```
Then it's just a matter of following the compiler:
```scala
private def modifyOrFail[E, C](f: B => Either[E, (C, A)])(implicit ea: EA <:< E, eb: EB <:< E): zio.IO[E, C] =
  original.modifyOrFail(b0 =>
    for {
      b <- getter(Right(b0))
      (c, a) <- f(b)
      a0 <- setter(Right(a))
    } yield (c, a0)
  )
```
Actually some future `recover*` combinator may change `setter` so I think this will need to be slightly modified to account for that.

Reply:
Thanks for the help will try that this weekend or next week

Reply:
@regiskuckaertz , @jdegoes I am having an issue with providing the right evidences: 
```
private def modifyOrFail[E, C](f: B => Either[E, (C, A)])(implicit ea: EA <:< E, eb: EB <:< E): zio.IO[E, C] =
  original.modifyOrFail(b0 =>
```
It's asking for the evidence of the original
```
def modifyOrFail[E, C](f: B0 => Either[E,(C, A0)])(implicit ea: EA0 <:< E, eb: EB0 <:< E): IO[E,C]
```
I have an `fea: EA0 => EA` and an `ea: EA <:< E` so it seems to me I should be able to get an 
`EA0 <:< E` , I havent figured out how though ...
And for `EB0 <:< E` I have no clue how to get a  hold of one.

Reply:
hi @dsebban and apologies for the late reply. I will find time this weekend to look into this in more detail and help you get this one over the line for 1.0 🎉